File: library/src/main/java/rxreddit/model/Link.java
package rxreddit.model;

import com.google.gson.annotations.SerializedName;

import java.util.List;

@SuppressWarnings("unused")
public class Link extends Listing implements Votable, Savable, Hideable {

    private Data data;

    public static class Data extends ListingData {

        @SerializedName("preview") Preview preview; // New field for preview images
        @SerializedName("domain") String domain;
        @SerializedName("banned_by") String bannedBy;
        @SerializedName("media_embed") MediaEmbed mediaEmbed;
        @SerializedName("subreddit") String subreddit;
        @SerializedName("selftext_html") String selftextHtml;
        @SerializedName("selftext") String selftext;
        @SerializedName("edited") String edited;
        @SerializedName("likes") Boolean likes;
        @SerializedName("user_reports") List<UserReport> userReports;
        @SerializedName("link_flair_text") String linkFlairText;
        @SerializedName("gilded") Integer gilded;
        @SerializedName("archived") Boolean isArchived;
        @SerializedName("clicked") Boolean clicked;
        @SerializedName("author") String author;
        @SerializedName("num_comments") Integer numComments;
        @SerializedName("score") Integer score;
        @SerializedName("approved_by") String approvedBy;
        @SerializedName("over_18") Boolean over18;
        @SerializedName("hidden") Boolean hidden;
        @SerializedName("thumbnail") String thumbnail;
        @SerializedName("subreddit_id") String subredditId;
        @SerializedName("hide_score") Boolean hideScore;
        @SerializedName("link_flair_css_class") String linkFlairCssClass;
        @SerializedName("author_flair_css_class") String authorFlairCssClass;
        @SerializedName("downs") Integer downs;
        @SerializedName("saved") Boolean saved;
        @SerializedName("stickied") Boolean stickied;
        @SerializedName("is_self") Boolean isSelf;
        @SerializedName("permalink") String permalink;
        @SerializedName("created") Double created;
        @SerializedName("url") String url;
        @SerializedName("author_flair_text") String authorFlairText;
        @SerializedName("title") String title;
        @SerializedName("created_utc") Double createdUtc;
        @SerializedName("distinguished") String distinguished;
        @SerializedName("media") Media media;
        @SerializedName("mod_reports") List<ModReport> modReports;
        @SerializedName("visited") Boolean visited;
        @SerializedName("num_reports") Integer numReports;
        @SerializedName("ups") Integer ups;
    }

    public static class Preview {

        @SerializedName("images") List<Image> images;

        public List<Image> getImages() {
            return images;
        }
    }

    //region Public API

    @Override
    public String getId() {
        return data.id;
    }

    public String getDomain() {
        return data.domain;
    }

    public MediaEmbed getMediaEmbed() {
        return data.mediaEmbed;
    }

    public String getSubreddit() {
        return data.subreddit;
    }

    public String getSelftextHtml() {
        return data.selftextHtml;
    }

    public String getSelftext() {
        return data.selftext;
    }

    public String isEdited() {
        return data.edited;
    }

    @Override
    public Boolean isLiked() {
        return data.likes;
    }

    @Override
    public void isLiked(Boolean b) {
        data.likes = b;
    }

    public List<UserReport> getUserReports() {
        return data.userReports;
    }

    public Object getLinkFlairText() {
        return data.linkFlairText;
    }

    public Integer getGilded() {
        return data.gilded;
    }

    public boolean isArchived() {
        return data.isArchived == null ? false : data.isArchived;
    }

    public boolean getClicked() {
        return data.clicked == null ? false : data.clicked;
    }

    public String getAuthor() {
        return data.author;
    }

    public Integer getNumComments() {
        return data.numComments;
    }

    public Integer getScore() {
        return data.score;
    }

    public void setScore(Integer score) {
        data.score = score;
    }

    @Override
    public void applyVote(int direction) {
        int scoreDiff = direction - getLikedScore();
        switch (direction) {
            case 0:
                isLiked(null);
                break;
            case 1:
                isLiked(true);
                break;
            case -1:
                isLiked(false);
                break;
        }
        if (data.score == null) return;
        data.score += scoreDiff;
    }

    private int getLikedScore() {
        if (isLiked() == null)
            return 0;
        else if (isLiked())
            return 1;
        else
            return -1;
    }

    public Object getApprovedBy() {
        return data.approvedBy;
    }

    public boolean getOver18() {
        return data.over18 == null ? false : data.over18;
    }

    public boolean isHidden() {
        return data.hidden == null ? false : data.hidden;
    }

    public void isHidden(Boolean b) {
        data.hidden = b;
    }

    public String getThumbnail() {
        return data.thumbnail;
    }

    public String getSubredditId() {
        return data.subredditId;
    }

    public boolean isScoreHidden() {
        return data.hideScore == null ? false : data.hideScore;
    }

    public Object getLinkFlairCssClass() {
        return data.linkFlairCssClass;
    }

    public Object getAuthorFlairCssClass() {
        return data.authorFlairCssClass;
    }

    public Integer getDowns() {
        return data.downs;
    }

    public boolean isSaved() {
        return data.saved == null ? false : data.saved;
    }

    public void isSaved(boolean b) {
        data.saved = b;
    }

    public boolean getStickied() {
        return data.stickied == null ? false : data.stickied;
    }

    public boolean isSelf() {
        return data.isSelf == null ? false : data.isSelf;
    }

    public String getPermalink() {
        return data.permalink;
    }

    public Double getCreated() {
        return data.created;
    }

    public String getUrl() {
        return data.url;
    }

    public Object getAuthorFlairText() {
        return data.authorFlairText;
    }

    public String getTitle() {
        return data.title;
    }

    public Double getCreatedUtc() {
        return data.createdUtc;
    }

    public String getDistinguished() {
        return data.distinguished;
    }

    public Media getMedia() {
        return data.media;
    }

    public List<ModReport> getModReports() {
        return data.modReports;
    }

    public boolean getVisited() {
        return data.visited == null ? false : data.visited;
    }

    public Object getNumReports() {
        return data.numReports;
    }

    public Integer getUps() {
        return data.ups;
    }

    public List<Image> getPreviewImages() {
        if (data.preview == null)
            return null;
        return data.preview.images;
    }

    //endregion
}

File: library/src/main/java/rxreddit/model/Link.kt
package rxreddit.model

import com.google.gson.annotations.SerializedName

data class Link(
    @SerializedName("data") val data: Data,
) : Listing(), Votable, Savable, Hideable {

    override val id: String
        get() = data.id

    val domain: String?
        get() = data.domain

    val mediaEmbed: MediaEmbed?
        get() = data.mediaEmbed

    val subreddit: String?
        get() = data.subreddit

    val selftextHtml: String?
        get() = data.selftextHtml

    val selftext: String?
        get() = data.selftext

    val edited: Boolean
        get() = data.edited ?: false

    override var liked: Boolean? by data::liked

    val userReports: List<UserReport>?
        get() = data.userReports

    val linkFlairText: Any?
        get() = data.linkFlairText

    val gilded: Int?
        get() = data.gilded

    override val archived: Boolean
        get() = data.isArchived ?: false

    val clicked: Boolean
        get() = data.clicked ?: false

    val author: String?
        get() = data.author

    val numComments: Int?
        get() = data.numComments

    var score: Int? by data::score

    override fun applyVote(direction: Int) {
        val scoreDiff = direction - likedScore
        data.liked = when (direction) {
            0 -> null
            1 -> true
            -1 -> false
            else -> null
        }
        if (data.score == null) return
        data.score = data.score?.plus(scoreDiff) ?: 0
    }

    private val likedScore: Int
        get() = if (liked == null) 0 else if (liked == true) 1 else -1

    val approvedBy: Any?
        get() = data.approvedBy

    val over18: Boolean
        get() = data.over18 ?: false

    override var hidden: Boolean
        get() = data.hidden ?: false
        set(value) {
            data.hidden = value
        }

    val thumbnail: String?
        get() = data.thumbnail

    val subredditId: String?
        get() = data.subredditId

    val isScoreHidden: Boolean
        get() = data.hideScore ?: false

    val linkFlairCssClass: Any?
        get() = data.linkFlairCssClass

    val authorFlairCssClass: Any?
        get() = data.authorFlairCssClass

    val downs: Int?
        get() = data.downs

    override var isSaved: Boolean
        get() = if (data.saved == null) false else data.saved!!
        set(value) {
            data.saved = value
        }

    val stickied: Boolean
        get() = data.stickied ?: false

    val isSelf: Boolean
        get() = data.isSelf ?: false

    val permalink: String?
        get() = data.permalink

    val created: Double?
        get() = data.created

    val url: String?
        get() = data.url

    val authorFlairText: Any?
        get() = data.authorFlairText

    val title: String?
        get() = data.title

    val createdUtc: Double?
        get() = data.createdUtc

    val distinguished: String?
        get() = data.distinguished

    val media: Media?
        get() = data.media

    val modReports: List<ModReport>?
        get() = data.modReports

    val visited: Boolean
        get() = data.visited ?: false

    val numReports: Any?
        get() = data.numReports

    val ups: Int?
        get() = data.ups

    val previewImages: List<Image>?
        get() = if (data.preview == null) null else data.preview!!.images

    data class Data(
        @SerializedName("preview")
        val preview: Preview? = null,

        @SerializedName("domain")
        val domain: String? = null,

        @SerializedName("banned_by")
        val bannedBy: String? = null,

        @SerializedName("media_embed")
        val mediaEmbed: MediaEmbed? = null,

        @SerializedName("subreddit")
        val subreddit: String? = null,

        @SerializedName("selftext_html")
        val selftextHtml: String? = null,

        @SerializedName("selftext")
        val selftext: String? = null,

        @SerializedName("edited")
        val edited: Boolean?,

        // TODO: Remove mutability here
        @SerializedName("likes")
        var liked: Boolean? = null,

        @SerializedName("user_reports")
        val userReports: List<UserReport>? = null,

        @SerializedName("link_flair_text")
        val linkFlairText: String? = null,

        @SerializedName("gilded")
        val gilded: Int? = null,

        @SerializedName("archived")
        val isArchived: Boolean? = null,

        @SerializedName("clicked")
        val clicked: Boolean? = null,

        @SerializedName("author")
        val author: String? = null,

        @SerializedName("num_comments")
        val numComments: Int? = null,

        // TODO: Remove mutability here
        @SerializedName("score")
        var score: Int? = null,

        @SerializedName("approved_by")
        val approvedBy: String? = null,

        @SerializedName("over_18")
        val over18: Boolean? = null,

        // TODO: Remove mutability here
        @SerializedName("hidden")
        var hidden: Boolean? = null,

        @SerializedName("thumbnail")
        val thumbnail: String? = null,

        @SerializedName("subreddit_id")
        val subredditId: String? = null,

        @SerializedName("hide_score")
        val hideScore: Boolean? = null,

        @SerializedName("link_flair_css_class")
        val linkFlairCssClass: String? = null,

        @SerializedName("author_flair_css_class")
        val authorFlairCssClass: String? = null,

        @SerializedName("downs")
        val downs: Int? = null,

        // TODO: Remove mutability here
        @SerializedName("saved")
        var saved: Boolean? = null,

        @SerializedName("stickied")
        val stickied: Boolean? = null,

        @SerializedName("is_self")
        val isSelf: Boolean? = null,

        @SerializedName("permalink")
        val permalink: String? = null,

        @SerializedName("created")
        val created: Double? = null,

        @SerializedName("url")
        val url: String? = null,

        @SerializedName("author_flair_text")
        val authorFlairText: String? = null,

        @SerializedName("title")
        val title: String? = null,

        @SerializedName("created_utc")
        val createdUtc: Double? = null,

        @SerializedName("distinguished")
        val distinguished: String? = null,

        @SerializedName("media")
        val media: Media? = null,

        @SerializedName("mod_reports")
        val modReports: List<ModReport>? = null,

        @SerializedName("visited")
        val visited: Boolean? = null,

        @SerializedName("num_reports")
        val numReports: Int? = null,

        @SerializedName("ups")
        val ups: Int? = null,

        @SerializedName("is_gallery")
        val isGallery: Boolean? = null,
    ) : ListingData()

    data class Preview(
        @SerializedName("images")
        val images: List<Image>? = emptyList(),
    )
}
